### PR TITLE
IGVF-51-publications

### PR DESCRIPTION
**Pull request by ian-whaling**

### CODE REVIEW

File: src/igvfd/schemas/changelogs/publication.md [hunk 1]
```diff
@@ -0,0 +1,3 @@
+## Changelog for publication.json
+
+### Minor changes since version 1
```
Comment (gabdank) [inline src/igvfd/schemas/changelogs/publication.md:3]: I don't think we should put the line about minor changes if there were no changes

File: src/igvfd/schemas/publication.json [hunk 2]
```diff
@@ -0,0 +1,135 @@
+{
+    "title": "Publication",
+    "description": "Schema for a publication page.",
```
Comment (gabdank) [inline src/igvfd/schemas/publication.json:3]: I think it is not publication page, but publication object

File: src/igvfd/schemas/publication.json [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+                "title": "Published by",
+                "type": "string",
+                "enum": [
+                    "community",
+                    "IGVF",
+                    "ENCODE"
+                ]
+            }
+        },
+        "publication_page": {
```
Comment (gabdank) [inline src/igvfd/schemas/publication.json:127]: I think we should drop it until we have pages editable by submitters

File: src/igvfd/schemas/publication.json [hunk 4]
```diff
@@ -0,0 +1,135 @@
+{
+    "title": "Publication",
+    "description": "Schema for a publication page.",
+    "$id": "/profiles/publication.json",
+    "$schema": "https://json-schema.org/draft/2020-12/schema",
+    "type": "object",
+    "required": [
+        "title",
```
Comment (gabdank) [inline src/igvfd/schemas/publication.json:8]: i think "identifiers" should be required

File: src/igvfd/schemas/publication.json [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+            "title": "Identifiers",
+            "description": "The identifiers that reference data found in the object.",
+            "type": "array",
+            "uniqueItems": true,
+            "items": {
+                "title": "Identifier",
+                "description": "An identifier that references data found in the object.",
+                "type": "string",
+                "uniqueKey": "publication:identifier",
+                "pattern": "^(PMID:[0-9]+|doi:10\\.[0-9]{4}[\\d\\s\\S\\:\\.\\/]+|PMCID:PMC[0-9]+|[0-9]{4}\\.[0-9]{4})$"
```
Comment (gabdank) [inline src/igvfd/schemas/publication.json:107]: I think the pattern should support whatever identifiers are in BioRxiv (and maybe it is already, just make sure)

File: src/igvfd/schemas/publication.json [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+            "type": "string"
+        },
+        "journal": {
+            "title": "Journal",
+            "description": "The journal of the publication.",
+            "type": "string"
+        },
+        "identifiers": {
+            "title": "Identifiers",
+            "description": "The identifiers that reference data found in the object.",
```
Comment (gabdank) [inline src/igvfd/schemas/publication.json:99]: I think the list of identifiers should have minItem=1, then the requirement to have this list would make sense

File: src/igvfd/tests/fixtures/schemas/publication.py [hunk 7]
```diff
@@ -0,0 +1,11 @@
+import pytest
+
+
+@pytest.fixture
+def publication(testapp, lab, award):
+    item = {
```
Comment (gabdank) [inline src/igvfd/tests/fixtures/schemas/publication.py:6]: if we make identifiers required will need to be updated

File: src/igvfd/tests/test_schema_publication.py [hunk 8]
```diff
@@ -0,0 +1,12 @@
+import pytest
+
+
+def test_date_published(publication, testapp):
+    res = testapp.patch_json(
```
Comment (gabdank) [inline src/igvfd/tests/test_schema_publication.py:5]: add test that makes sure identifiers are unique

File: src/igvfd/tests/test_schema_publication.py [hunk 9]
```diff
@@ -0,0 +1,12 @@
...
+        publication['@id'],
+        {'date_published': '2022-2-2'}, expect_errors=True)
```
Comment (gabdank) [inline src/igvfd/tests/test_schema_publication.py:7]: add test for required properties
and test for regex of identifiers

**Review by gabdank — CHANGES_REQUESTED** [review]

File: src/igvfd/schemas/publication.json (hunk 5 above)
Comment (ian-whaling) [reply src/igvfd/schemas/publication.json:109]: BioRxiv uses doi, so it is supported. I put an insert to reflect this.

File: src/igvfd/types/publication.py [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+        keys = super(Publication, self).unique_keys(properties)
+        if properties.get('identifiers'):
+            keys.setdefault('alias', []).extend(properties['identifiers'])
+        return keys
+
+    @calculated_property(condition='date_published', schema={
+        'title': 'Publication year',
+        'type': 'integer',
+    }
+    )
```
Comment (gabdank) [inline src/igvfd/types/publication.py:33]: please put these two on the same line like line 18

File: src/igvfd/schemas/publication.json [hunk 11]
```diff
@@ -0,0 +1,131 @@
...
+                    "format": "date"
+                },
+                {
+                    "comment": "Historical date formats, not valid for new submissions.",
+                    "permission": "import_items",
+                    "pattern": "^[0-9]{4}[^-]*$"
+                }
+            ]
+        },
+        "date_revised": {
```
Comment (gabdank) [inline src/igvfd/schemas/publication.json:66]: we discussed it at a wranglers meeting, do not use  date-time for any of the date properties

only timestamps are date-time

File: src/igvfd/schemas/publication.json [hunk 12]
```diff
@@ -0,0 +1,131 @@
...
+            "type": "string",
+            "oneOf": [
+                {
+                    "format": "date"
+                },
+                {
+                    "comment": "Historical date formats, not valid for new submissions.",
+                    "permission": "import_items",
+                    "pattern": "^[0-9]{4}[^-]*$"
+                }
```
Comment (gabdank) [inline src/igvfd/schemas/publication.json:63]: what is the purpose of these?

File: src/igvfd/types/publication.py [hunk 13]
```diff
@@ -0,0 +1,39 @@
...
+    def publication_year(self, date_published):
+        likely_year = date_published[:4]
```
Comment (gabdank) [inline src/igvfd/types/publication.py:35]: I think you should use python date conversions, something like: 
```
dt = datetime.strptime(datestring, "%Y-%m-%d")
print (dt.year)
```

**Review by gabdank — CHANGES_REQUESTED** [review]

File: src/igvfd/schemas/publication.json (hunk 12 above)
Comment (ian-whaling) [reply src/igvfd/schemas/publication.json:63]: Not sure, this was ported from ENCODE but it looks to have been deprecated there so I will remove it.

File: src/igvfd/tests/test_types_publication.py [hunk 14]
```diff
@@ -0,0 +1,8 @@
+import pytest
+
+
+def test_publication_year(testapp, publication):
```
Comment (gabdank) [inline src/igvfd/tests/test_types_publication.py:4]: can you please add a test to make sure the publication without date_published works ok?

**Review by gabdank — CHANGES_REQUESTED** [review]

File: src/igvfd/types/publication.py [hunk 15]
```diff
@@ -0,0 +1,36 @@
...
+    datetime
+)
+
+
+@collection(
+    name='publications',
+    unique_key='publication:identifier',
+    properties={
+        'title': 'Publications',
+        'description': 'Publication pages',
```
Comment (gabdank) [inline src/igvfd/types/publication.py:18]: I don't think the description should be "Publication pages", should not it be "Listing of publications"?

**Review by gabdank — CHANGES_REQUESTED** [review]
please update the description of the collection of publications. Otherwise, I think it is good to go.

**Review by gabdank — APPROVED** [review]

**Pull request merged**